File: Tyche.iOS/Session/Sources/Session/Authentication/Infrastructure/AuthenticationFirebaseDataSource.swift
import FirebaseAuth

class AuthenticationFirebaseDataSource: AuthenticationExternalDataSource {
    private let firebaseAuth: Auth

    private let signInUrlTemplate: SignInLinkUrlTemplateProvider

    init(firebaseAuth: Auth, signInUrlTemplate: SignInLinkUrlTemplateProvider) {
        self.firebaseAuth = firebaseAuth
        self.signInUrlTemplate = signInUrlTemplate
    }
    
    func sendSignInLinkToEmail(email: String) async throws {
        let actionCodeSettings = ActionCodeSettings()
        actionCodeSettings.url = URL(string: String(format: signInUrlTemplate(), email))
        actionCodeSettings.handleCodeInApp = true
        try await firebaseAuth.sendSignInLink(toEmail: email, actionCodeSettings: actionCodeSettings)
    }
    
    func signInWithEmailLink(email: String, emailLink: String) async throws -> ExternalAccountId {
        let authResult = try await firebaseAuth.signIn(withEmail: email, link: emailLink)
        return authResult.user.uid
    }

    func signInWithEmailAndPassword(email: String, password: String) async throws -> ExternalAccountId {
        let authResult = try await firebaseAuth.signIn(withEmail: email, password: password)
        return authResult.user.uid
    }

    func isSignInWithEmailLink(emailLink: String) async -> Bool {
        firebaseAuth.isSignIn(withEmailLink: emailLink)
    }
    
    func signOut() async throws {
        try firebaseAuth.signOut()
    }
}

File: Android/session/src/main/java/com/felipearpa/tyche/session/authentication/infrastructure/AuthenticationFirebaseDataSource.kt
package com.felipearpa.tyche.session.authentication.infrastructure

import android.content.Context
import com.felipearpa.tyche.core.IosBundleIdProvider
import com.felipearpa.tyche.session.SignInLinkUrlTemplateProvider
import com.felipearpa.tyche.session.authentication.domain.AuthenticationExternalDataSource
import com.felipearpa.tyche.session.authentication.domain.ExternalAccountId
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.actionCodeSettings
import kotlinx.coroutines.tasks.await

internal class AuthenticationFirebaseDataSource(
    private val firebaseAuth: FirebaseAuth,
    private val signInLinkUrlTemplate: SignInLinkUrlTemplateProvider,
    private val iosBundleId: IosBundleIdProvider,
    private val context: Context,
) : AuthenticationExternalDataSource {

    override suspend fun sendSignInLinkToEmail(email: String) {
        val actionCodeSettings = actionCodeSettings {
            url = String.format(signInLinkUrlTemplate(), email)
            handleCodeInApp = true
            setAndroidPackageName(context.packageName, true, null)
            iosBundleId = iosBundleId()
        }
        firebaseAuth.sendSignInLinkToEmail(email, actionCodeSettings).await()
    }

    override suspend fun signInWithEmailLink(email: String, emailLink: String): ExternalAccountId {
        val authResult = firebaseAuth.signInWithEmailLink(email, emailLink).await()
        return authResult.user!!.uid
    }

    override suspend fun signInWithEmailAndPassword(
        email: String,
        password: String,
    ): ExternalAccountId {
        val authResult = firebaseAuth.signInWithEmailAndPassword(email, password).await()
        return authResult.user!!.uid
    }

    override suspend fun isSignInWithEmailLink(emailLink: String): Boolean =
        firebaseAuth.isSignInWithEmailLink(emailLink)

    override suspend fun signOut() {
        firebaseAuth.signOut()
    }
}
